File: lint-rules-android-lint/src/test/java/com/vanniktech/lintrules/android/DefaultLayoutAttributeDetectorTest.java
package com.vanniktech.lintrules.android;

import com.android.tools.lint.checks.infrastructure.LintDetectorTest;
import com.android.tools.lint.detector.api.Detector;
import com.android.tools.lint.detector.api.Issue;
import java.util.List;
import org.intellij.lang.annotations.Language;

import static com.vanniktech.lintrules.android.AndroidDetectorTest.NO_WARNINGS;
import static com.vanniktech.lintrules.android.DefaultLayoutAttributeDetector.ISSUE_DEFAULT_LAYOUT_ATTRIBUTE;
import static java.util.Collections.singletonList;
import static org.fest.assertions.api.Assertions.assertThat;

public final class DefaultLayoutAttributeDetectorTest extends LintDetectorTest {
  public void testTextStyleNormal() throws Exception {
    @Language("XML") final String source = "<TextView\n"
        + "  xmlns:android=\"http://schemas.android.com/apk/res/android\"\n"
        + "  android:layout_width=\"wrap_content\"\n"
        + "  android:layout_height=\"wrap_content\"\n"
        + "  android:textStyle=\"normal\"\n"
        + "  />";

    assertThat(lintProject(xml("/res/layout/ids.xml", source))).isEqualTo("res/layout/ids.xml:5: Warning: textStyle=\"normal\" is the default and hence you don't need to specify it. [DefaultLayoutAttribute]\n"
        + "  android:textStyle=\"normal\"\n"
        + "                     ~~~~~~\n"
        + "0 errors, 1 warnings\n");
  }

  public void testTextStyleBold() throws Exception {
    @Language("XML") final String source = "<TextView\n"
        + "  xmlns:android=\"http://schemas.android.com/apk/res/android\"\n"
        + "  android:layout_width=\"wrap_content\"\n"
        + "  android:layout_height=\"wrap_content\"\n"
        + "  android:textStyle=\"bold\"\n"
        + "  />";

    assertThat(lintProject(xml("/res/layout/ids.xml", source))).isEqualTo(NO_WARNINGS);
  }

  public void testTextStyleNormalIgnored() throws Exception {
    @Language("XML") final String source = "<TextView\n"
        + "  xmlns:android=\"http://schemas.android.com/apk/res/android\"\n"
        + "  xmlns:tools=\"http://schemas.android.com/tools\"\n"
        + "  android:layout_width=\"wrap_content\"\n"
        + "  android:layout_height=\"wrap_content\"\n"
        + "  android:textStyle=\"normal\"\n"
        + "  tools:ignore=\"DefaultLayoutAttribute\"\n"
        + "  />";

    assertThat(lintProject(xml("/res/layout/ids.xml", source))).isEqualTo(NO_WARNINGS);
  }

  public void testShouldNotCrashWithStyle() throws Exception {
    @Language("XML") final String source = "<TextView\n"
        + "      style=\"?android:attr/borderlessButtonStyle\"\n"
        + "      />";

    assertThat(lintProject(xml("/res/layout/ids.xml", source))).isEqualTo(NO_WARNINGS);
  }

  @Override protected Detector getDetector() {
    return new DefaultLayoutAttributeDetector();
  }

  @Override protected List<Issue> getIssues() {
    return singletonList(ISSUE_DEFAULT_LAYOUT_ATTRIBUTE);
  }

  @Override protected boolean allowCompilationErrors() {
    return false;
  }
}

File: lint-rules-android-lint/src/test/java/com/vanniktech/lintrules/android/DefaultLayoutAttributeDetectorTest.kt
package com.vanniktech.lintrules.android

import com.android.tools.lint.checks.infrastructure.TestFiles.xml
import com.android.tools.lint.checks.infrastructure.TestLintTask.lint
import org.junit.Test

class DefaultLayoutAttributeDetectorTest {
  @Test fun textStyleNormal() {
    lint()
        .files(xml("res/layout/ids.xml", """
          |<TextView
          |    xmlns:android="http://schemas.android.com/apk/res/android"
          |    android:layout_width="wrap_content"
          |    android:layout_height="wrap_content"
          |    android:textStyle="normal"/>""".trimMargin()))
        .issues(ISSUE_DEFAULT_LAYOUT_ATTRIBUTE)
        .run()
        .expect("""
          |res/layout/ids.xml:5: Warning: This is the default and hence you don't need to specify it. [DefaultLayoutAttribute]
          |    android:textStyle="normal"/>
          |                       ~~~~~~
          |0 errors, 1 warnings""".trimMargin())
        .expectFixDiffs("""
          |Fix for res/layout/ids.xml line 4: Remove:
          |@@ -4 +4
          |-     android:layout_height="wrap_content"
          |-     android:textStyle="normal" />
          |@@ -6 +4
          |+     android:layout_height="wrap_content" />
          |""".trimMargin())
  }

  @Test fun textStyleBold() {
    lint()
        .files(xml("res/layout/ids.xml", """
          |<TextView
          |    xmlns:android="http://schemas.android.com/apk/res/android"
          |    android:layout_width="wrap_content"
          |    android:layout_height="wrap_content"
          |    android:textStyle="bold"/>""".trimMargin()))
        .issues(ISSUE_DEFAULT_LAYOUT_ATTRIBUTE)
        .run()
        .expectClean()
  }

  @Test fun textStyleNormalIgnored() {
    lint()
        .files(xml("res/layout/ids.xml", """
          |<TextView
          |    xmlns:android="http://schemas.android.com/apk/res/android"
          |    xmlns:tools="http://schemas.android.com/tools
          |    android:layout_width="wrap_content"
          |    android:layout_height="wrap_content"
          |    android:textStyle="normal"
          |    tools:ignore="DefaultLayoutAttribute"/>""".trimMargin()))
        .issues(ISSUE_DEFAULT_LAYOUT_ATTRIBUTE)
        .run()
        .expectClean()
  }

  @Test fun shouldNotCrashWithStyle() {
    lint()
        .files(xml("res/layout/ids.xml", """
          |<TextView
          |    style="?android:attr/borderlessButtonStyle"/>""".trimMargin()))
        .issues(ISSUE_DEFAULT_LAYOUT_ATTRIBUTE)
        .run()
        .expectClean()
  }
}
